### PR TITLE
feat: add LifecyclePreLoad

### PR DESCRIPTION
**Pull request by akitaSummer**
##### Checklist

- [ ] `npm test` passes
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message follows commit guidelines

## Summary by CodeRabbit

- **New Features**
	- Introduced static lifecycle hooks for improved lifecycle management.
	- Added a `preLoad` method to enhance egg object initialization.
	- Implemented a `preLoad` method for module pre-loading operations.
	- Created a new class `Foo` to manage lifecycle events with various hooks.

- **Bug Fixes**
	- Enhanced error handling in the `preLoad` function to provide clearer error messages.

- **Tests**
	- Expanded test suite to validate the functionality of the new lifecycle methods and pre-loading behavior.

- **Documentation**
	- Added metadata for the new lifecycle module in `package.json`.

### CODE REVIEW

File: core/types/lifecycle/EggObjectLifecycle.ts [hunk 1]
```diff
@@ -4,6 +4,10 @@ import type { EggObject, EggObjectLifeCycleContext } from '@eggjs/tegg-runtime';
  * lifecycle hook interface for egg object
  */
 export interface EggObjectLifecycle {
+  /**
+   * call before standalone load
+   */
+  preStandaloneLoad?(ctx: EggObjectLifeCycleContext, eggObj: EggObject): Promise<void>;
```
Comment (killagu) [inline core/types/lifecycle/EggObjectLifecycle.ts:10]: preload 和 standalone 应该没有强关联。

File: standalone/standalone/test/index.test.ts [hunk 2]
```diff
@@ -276,4 +276,26 @@ describe('standalone/standalone/test/index.test.ts', () => {
...
+    it('preLoad should work', async () => {
+      await preLoad(fixturePath);
+      // eslint-disable-next-line @typescript-eslint/no-var-requires
+      const { Foo } = require(path.join(fixturePath, './foo'));
+      deepStrictEqual(Foo.staticCalled, [ 'preStandaloneLoad' ]);
+    });
+
+    it('should work', async () => {
+      const result = await main<string[]>(fixturePath);
+      assert.deepStrictEqual(result, [
```
Comment (killagu) [inline standalone/standalone/test/index.test.ts:292]: 这里应该要有 preload 的？

File: standalone/standalone/src/Runner.ts [hunk 3]
```diff
@@ -189,6 +191,24 @@ export class Runner {
...
+    for (const module of moduleReferences) {
+      const protoClassList = LoaderFactory.createLoader(module.path, 'MODULE').load();
+      // lifecycle preLoad hook
+      for (const protoClass of protoClassList) {
+        const fnName = LifecycleUtil.getStaticLifecycleHook('preLoad', protoClass);
+        if (fnName) {
+          await protoClass[fnName]();
+        }
+      }
+    }
```
Comment (killagu) [inline standalone/standalone/src/Runner.ts:209]: 这个是否放在 LoadUnit 里做更加合适？

File: standalone/standalone/src/Runner.ts [hunk 4]
```diff
@@ -189,6 +191,24 @@ export class Runner {
     return [ standaloneLoadUnit, ...loadUnits ];
   }
 
+  static async preLoad(cwd: string, dependencies?: RunnerOptions['dependencies']) {
+    const moduleDirs = (dependencies || []).concat(cwd);
+    const moduleReferences = moduleDirs.reduce((list, baseDir) => {
+      const module = typeof baseDir === 'string' ? { baseDir } : baseDir;
+      return list.concat(...ModuleConfigUtil.readModuleReference(module.baseDir, module));
+    }, [] as readonly ModuleReference[]);
```
Comment (killagu) [inline standalone/standalone/src/Runner.ts:199]: 这里和 constructor 重复了，是否需要有通用的方法?

File: standalone/standalone/src/EggModuleLoader.ts [hunk 5]
```diff
@@ -10,9 +10,9 @@ export class EggModuleLoader {
     this.moduleReferences = moduleReferences;
   }
 
-  private buildAppGraph(loaderCache: Map<string, Loader>) {
+  private static buildAppGraph(loaderCache: Map<string, Loader>, moduleReferences: readonly ModuleReference[]) {
```
Comment (killagu) [inline standalone/standalone/src/EggModuleLoader.ts:13]: 这样改了是 break change。不应该这么改。

Comment (akitaSummer) [reply standalone/standalone/src/EggModuleLoader.ts]: buildAppGraph 作为 private 函数仅在内部使用，不会影响外部，已改

**Review by killagu — APPROVED** [review]
LGTM

**Comment by killagu** [comment]
```
 - @eggjs/ajv-decorator@3.42.0
 - @eggjs/aop-decorator@3.42.0
 - @eggjs/controller-decorator@3.42.0
 - @eggjs/core-decorator@3.42.0
 - @eggjs/dal-decorator@3.42.0
 - @eggjs/dal-runtime@3.42.0
 - @eggjs/egg-module-common@3.42.0
 - @eggjs/eventbus-decorator@3.42.0
 - @eggjs/standalone-decorator@3.42.0
 - @eggjs/tegg@3.42.0
 - @eggjs/tegg-ajv-plugin@3.42.0
 - @eggjs/tegg-aop-plugin@3.42.0
 - @eggjs/tegg-aop-runtime@3.42.0
 - @eggjs/tegg-background-task@3.42.0
 - @eggjs/tegg-common-util@3.42.0
 - @eggjs/tegg-config@3.42.0
 - @eggjs/tegg-controller-plugin@3.42.0
 - @eggjs/tegg-dal-plugin@3.42.0
 - @eggjs/tegg-dynamic-inject@3.42.0
 - @eggjs/tegg-dynamic-inject-runtime@3.42.0
 - @eggjs/tegg-eventbus-plugin@3.42.0
 - @eggjs/tegg-eventbus-runtime@3.42.0
 - @eggjs/tegg-lifecycle@3.42.0
 - @eggjs/tegg-loader@3.42.0
 - @eggjs/tegg-metadata@3.42.0
 - @eggjs/tegg-orm-decorator@3.42.0
 - @eggjs/tegg-orm-plugin@3.42.0
 - @eggjs/tegg-plugin@3.42.0
 - @eggjs/tegg-runtime@3.42.0
 - @eggjs/tegg-schedule-decorator@3.42.0
 - @eggjs/tegg-schedule-plugin@3.42.0
 - @eggjs/tegg-standalone@3.42.0
 - @eggjs/tegg-transaction-decorator@3.42.0
 - @eggjs/tegg-types@3.42.0
 ```

**Pull request merged**